contracts/contracts/GrantRoundManager.sol

```diff
@@ -32,35 +24,70 @@ contract GrantRoundManager {
   /// @notice Address of the ERC20 token in which donations are made
   IERC20 public immutable donationToken;
 
+  /// @notice Used during donations to temporarily store swap output amounts
+  mapping(IERC20 => uint256) internal swapOutputs;
+
+  /// @notice Used during donations to temporarily store donation ratio sums to ensure totals are always 100%
+  mapping(IERC20 => uint256) internal donationRatios;
+
   /// @notice WETH address
-  address public constant WETH = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+  IERC20 public constant WETH = IERC20(0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2);
+
+  /// @notice Scale factor
+  uint256 internal constant WAD = 1e18;
+
+  /// --- Types ---
+  /// @notice Defines the total `amount` of the specified `token` that needs to be swapped to `donationToken`. If
+  /// `path == donationToken`, no swap is required and we just transfer the tokens
+  struct SwapSummary {
+    uint256 amountIn;
+    uint256 amountOutMin; // minimum amount to be returned after swap
+    bytes path;
+  }
 
+  /// @notice Donation inputs and Uniswap V3 swap inputs: https://docs.uniswap.org/protocol/guides/swaps/multihop-swaps
+  struct Donation {
+    uint96 grantId; // grant ID to which donation is being made
+    IERC20 token; // address of the token to donate
+    uint256 ratio; // ratio of `token` to donate, specified as numerator where WAD = 1e18 = 100%
+    GrantRound[] rounds; // rounds against which the donation should be counted
+  }
+
+  // --- Events ---
   /// @notice Emitted when a new GrantRound contract is created
   event GrantRoundCreated(address grantRound);
 
   /// @notice Emitted when a donation has been made
-  event GrantDonation(
-    uint96 indexed grantId,
-    IERC20 indexed tokenIn,
-    uint256 amountIn,
-    uint256 amountOut,
-    GrantRound[] rounds
-  );
+  event GrantDonation(uint96 indexed grantId, IERC20 indexed tokenIn, uint256 donationAmount, GrantRound[] rounds);
 
+  // --- Constructor ---
   constructor(
     GrantRegistry _registry,
     ISwapRouter _router,
     IERC20 _donationToken
   ) {
+    // Validation
     require(_registry.grantCount() >= 0, "GrantRoundManager: Invalid registry");
     require(address(_router).isContract(), "GrantRoundManager: Invalid router"); // Router interface doesn't have a state variable to check
     require(_donationToken.totalSupply() > 0, "GrantRoundManager: Invalid token");
 
+    // Set state
     registry = _registry;
     router = _router;
     donationToken = _donationToken;
+
+    // Token approvals of common tokens
+    // TODO inherit from SwapRouter to remove the need for this approvals and extra safeTransferFrom before swap
+    IERC20(0x6B175474E89094C44Da98b954EedeAC495271d0F).safeApprove(address(_router), type(uint256).max); // DAI
```

Comment:
Just want to make sure I understand: as currently written, the donate method will fail if it's not using one of these pre-approved tokens, and there is no mechanism for the contract to approve new ones. Correct?

Inheriting from SwapRouter means we'd become a custom v3 router, and users would have to trust (or verify by inspected verified source code) that we did not modify the router functionality in some malicious or unsafe way. Correct?

Reply:
Correct on all counts! If for whatever reason we don't end up inheriting from SwapRouter, I think we'd want a public method where you pass in a token address and it approves the SwapRouter to spend MaxUint of that token. Otherwise we have to check allowances every donation which would get very costly

Reply:
Got it, sounds good. I think we want to go this route (ha!). Created this issue to track as a separate PR: https://github.com/dcgtc/dgrants/issues/99

Reply:
Just a clarification -> Even if we inherited the list -> wouldn't we still have to do an approve for every token on the list right ? (the only diff is we won't be hardcoding every token within this contract )

Reply:
So here is the flow when you normally use Uniswap as a user:
1. Approve router to spend your tokens
2. Call swap on router, which takes your tokens and executes swap 
3. (1 total approval, 1 transferFrom)

Here is our current flow:
1. Approve manager to spend your tokens
2. Manager approves router to spend it's tokens
3. Call donate on manager, which takes your tokens, the router takes tokens from manager to execute swap (no more approvals)
4. (2 total approvals, 2 transferFroms)

So once we inherit from SwapRouter, we'll migrate to the top flow, which will reduce approvals and transfers